daemon/logger/jsonfilelog/read.go

```diff
@@ -105,8 +110,7 @@ func (d *decoder) Decode() (msg *logger.Message, err error) {
 		// If the json logger writes a partial json log entry to the disk
 		// while at the same time the decoder tries to decode it, the race condition happens.
 		if err == io.ErrUnexpectedEOF {
-			d.rdr = io.MultiReader(d.dec.Buffered(), d.rdr)
-			d.dec = json.NewDecoder(d.rdr)
+			d.dec = json.NewDecoder(io.MultiReader(d.dec.Buffered(), d.rdr))
```

Comment:
iiuc the issue in here is not the resetting `d.rdr` or not (what looks more correct to me in the old version) but that we use incompatible `Reader` where after it returns `io.EOF` we can read again and it may return more data then. This does not follow the `io.Reader` interface that stdlib defines and therefore the utility functions behavior gets wrong. More specifically `io.MultiReader` caches `io.EOF` https://github.com/golang/go/blob/414fa8c35e7c2f65e2c767d6db2f25791e53b5c1/src/io/multi.go#L41 . So in the previous code if the `MultiReader` has already returned `EOF` once and now is added as a second reader to a new `MultiReader` it never gets called again and new `MultiReader` returns `EOF` without making any new `Read` calls to the original `d.rdr`.

Reply:
Added a 2nd commit per your suggestion to replace io.MultiReader with a custom implementation that does not cache EOF's so that we can keep the full reader (with the buffered data) in `d.rdr`.